vector/src/main/java/im/vector/LoginHandler.java

```diff
@@ -22,87 +22,155 @@
+            @Override
+            public void onSuccess(Credentials credentials) {
+                // sanity check - GA issue
+                if (TextUtils.isEmpty(credentials.userId)) {
+                    callback.onMatrixError(new MatrixError(MatrixError.FORBIDDEN, "No user id"));
+                    return;
+                }
+
+                hsConfig.setCredentials(credentials);
+
+                // Check whether a shadow HS is available in the Tchap configuration
+                final HomeServerConnectionConfig shadowHS = tchapConfig.getShadowHSConfig();
+                if (shadowHS != null) {
+                    shadowLogin(context, shadowHS, username, phoneNumber, phoneNumberCountry, password, new SuccessCallback<String>() {
+                        @Override
+                        public void onSuccess(String warningMessage) {
+                            onRegistrationDone(Matrix.getInstance(context.getApplicationContext()), tchapConfig);
+                            callback.onSuccess(warningMessage);
+                        }
+                    });
+
+                } else {
+                    onRegistrationDone(Matrix.getInstance(context.getApplicationContext()), tchapConfig);
+                    callback.onSuccess(null);
+                }
+            }
+
+            @Override
+            public void onAcceptedCert() {
+                login(context, tchapConfig, username, phoneNumber, phoneNumberCountry, password, callback);
+            }
+        });
+    }
+
+    /**
+     * Handle the login stage on the shadow HS.
+     * This method returns a warning message in case of failure, otherwise it returns null.
+     *
+     * @param context            the context.
+     * @param shadowHSConfig     The shadow homeserver config.
+     * @param username           The username.
+     * @param phoneNumber        The phone number.
+     * @param phoneNumberCountry The phone number country code.
+     * @param password           The password;
+     * @param callback           The callback.
+     */
+    private void shadowLogin(final Context context,
+                             final HomeServerConnectionConfig shadowHSConfig,
+                             final String username,
+                             final String phoneNumber,
+                             final String phoneNumberCountry,
+                             final String password,
+                             final SuccessCallback<String> callback) {
+        callLogin(context, shadowHSConfig, username, phoneNumber, phoneNumberCountry, password, new UnrecognizedCertApiCallback<Credentials>(shadowHSConfig) {
+            private void onError(String errorMessage) {
+                Log.e(LOG_TAG, "Login to the shadow HS failed " + errorMessage);
+                callback.onSuccess(context.getString(R.string.tchap_auth_agent_failure_warning_msg));
```

Comment:
It's strange to call `onSuccess()` for an error, but I understand that it is only a warning?

Reply:
yes it is only a warning, the main registration must not failed in case of failure during shadow login

Reply:
FYI I updated the code by using directly a MediaScanDao instance

Reply:
MediaScanDao?

Reply:
this comment about " MediaScanDao" is not a the right place